Feature: custom modules (name, icon, command)

clotodex opened the pull request:
- config parser to allow multiple custom modules that are parsed based on their name
- custom module that will render configurable nerdfont icon
- custom command that gets executed on click

```toml
[modules]
# The modules that will be displayed on the left side of the status bar
left = [
	"appLauncher",
    [ "qr-scan", "updates"],
	"windowTitle",
]
center = [ "workspaces" ]
right = [ "qr-scan", "clipboard", "systemInfo", "tray", [ "clock", "privacy", "settings" ] ]

[[CustomModule]]
name = "qr-scan"
icon = "󰐲"
command = "screenshot-area-scan-qr"
```

To be discussed:
To parse the config I ran into serdes limitation of not being able to parse nested enums with different taggings.
So instead I am now parsing the ModuleNames directly from string - this also has benefits, like e.g. allowing any spelling to map to the module.

MalpenZibo commented:
Ok, I saw the code, and it seems very promising.

The problem with serde seems related to the `toml` crate.
I tried with this enum
```
#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CustomModule {
    pub icon: String,
    pub command: String,
}

#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ModuleName {
    AppLauncher,
    Updates,
    Clipboard,
    Workspaces,
    WindowTitle,
    SystemInfo,
    KeyboardLayout,
    KeyboardSubmap,
    Tray,
    Clock,
    Privacy,
    Settings,
    MediaPlayer,
    CustomModule(CustomModule),
}
```

With `json_serde`, it seems to be working fine. I need to understand if I like your solution more or this one.

Your solution has some advantages. Starting with the enum naming, `ModuleName` indicates only the module names and not the module config data. For this reason, maybe splitting the two things could be better. It would be possible to place the same module in multiple locations without redefining the config (I don't think it's something useful, but right now it could be done with every module).

The point is:
I like your solution, but I don't like this change (and everything related) too much
```
pub enum ModuleDef {
    Single(String),
    Group(Vec<String>),
}
```

I prefer using the ModuleName. I know that's a limitation of toml_serde, and for this reason, I'm thinking of switching to json_serde. I haven't released a version with the toml config, so it could be THE moment to make this change. I know everyone using the `dev` version will hate me sooo much :sweat_smile: 

What do you think?

clotodex commented:
Yeah this was my going back and forth when setting it up

In the grand scheme of things i am using nixos so i will likely anyways build my own nix based module in the end so please dont follow my preferences here

Why i like the toml method is:
- toml overall has a really nice modular design, especially the custom modules
- json just works but is in my eyes a quite clunky config format

Why i think toml is maintainable:

I think i would throw out the auto-parsing here and parse ALL modules as strings 
Then it is as "simple" as before, no big code duplication and stays "single & group" - simply would not auto-parse. However we can easily give then also custom error messages like "module not defined".

These are my two cents - hope it helps

clotodex commented:
ALSO I already implemented dynamic custom modules including an alert system (just not pr ready yet)
This allows you to specify a command that yields updates and can change text and icon and alert status (red dot) dynamically

https://github.com/clotodex/ashell-iwd/tree/feature%2Fdynamic-custom-components

MalpenZibo commented on src/config.rs:
```diff
@@ -348,14 +348,44 @@ pub enum ModuleName {
+    Custom(String),
+}
+
+impl From<String> for ModuleName {
+    fn from(value: String) -> Self {
+        // TODO: this could also be parsed by a deserialize helper { module: ModuleName } (direct
+        // from string is not valid toml)
+        // Benefit of this appraoch is we could match by lowercase and support all spellings
+        match value.as_str() {
+            "appLauncher" => ModuleName::AppLauncher,
+            "updates" => ModuleName::Updates,
+            "clipboard" => ModuleName::Clipboard,
+            "workspaces" => ModuleName::Workspaces,
+            "windowTitle" => ModuleName::WindowTitle,
+            "systemInfo" => ModuleName::SystemInfo,
+            "keyboardLayout" => ModuleName::KeyboardLayout,
+            "keyboardSubmap" => ModuleName::KeyboardSubmap,
+            "tray" => ModuleName::Tray,
+            "clock" => ModuleName::Clock,
+            "privacy" => ModuleName::Privacy,
+            "settings" => ModuleName::Settings,
+            "mediaPlayer" => ModuleName::MediaPlayer,
+            _ => ModuleName::Custom(value),
+        }
+    }
+}
+impl<'a> From<&'a str> for ModuleName {
+    fn from(value: &'a str) -> ModuleName {
+        String::into(value.to_owned())
+    }
 }
```
Let's try this. It's a custom deserializer. Tries to deserialize every entry in a ModuleName, and in case there isn't a match it uses the Custom ModuleName

```suggestion
impl<'de> Deserialize<'de> for ModuleName {
    fn deserialize<D>(deserializer: D) -> Result<ModuleName, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct ModuleNameVisitor;

        impl Visitor<'_> for ModuleNameVisitor {
            type Value = ModuleName;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("a camelCase string representing a ModuleName")
            }

            fn visit_str<E>(self, value: &str) -> Result<ModuleName, E>
            where
                E: Error,
            {
                Ok(match value {
                    "appLauncher" => ModuleName::AppLauncher,
                    "updates" => ModuleName::Updates,
                    "clipboard" => ModuleName::Clipboard,
                    "workspaces" => ModuleName::Workspaces,
                    "windowTitle" => ModuleName::WindowTitle,
                    "systemInfo" => ModuleName::SystemInfo,
                    "keyboardLayout" => ModuleName::KeyboardLayout,
                    "keyboardSubmap" => ModuleName::KeyboardSubmap,
                    "tray" => ModuleName::Tray,
                    "clock" => ModuleName::Clock,
                    "privacy" => ModuleName::Privacy,
                    "settings" => ModuleName::Settings,
                    "mediaPlayer" => ModuleName::MediaPlayer,
                    other => ModuleName::Custom(other.to_string()),
                })
            }
        }

        deserializer.deserialize_str(ModuleNameVisitor)
    }
}
```

MalpenZibo commented on src/config.rs:
```diff
@@ -348,14 +348,44 @@ pub enum ModuleName {
 pub enum ModuleDef {
-    Single(ModuleName),
-    Group(Vec<ModuleName>),
+    Single(String),
+    Group(Vec<String>),
```
This could return to be a ModuleName

MalpenZibo commented on src/modules/mod.rs:
```diff
@@ -59,8 +62,18 @@ impl App {
-                ModuleDef::Group(group) => self.group_module_wrapper(group, id, opacity),
+                // life parsing of string to module
+                ModuleDef::Single(module) => {
+                    self.single_module_wrapper(module.to_owned().into(), id, opacity)
```
```suggestion
                    self.single_module_wrapper(module, id, opacity)
```

MalpenZibo commented on src/modules/mod.rs:
```diff
@@ -59,8 +62,18 @@ impl App {
+                ModuleDef::Single(module) => {
+                    self.single_module_wrapper(module.to_owned().into(), id, opacity)
+                }
+                ModuleDef::Group(group) => self.group_module_wrapper(
+                    &group
+                        .iter()
+                        .map(|m| m.to_owned().into())
+                        .collect::<Vec<_>>(),
+                    id,
+                    opacity,
+                ),
```
```suggestion
                ModuleDef::Group(group) => self.group_module_wrapper(
                    &group,
                    id,
                    opacity,
                ),
```

MalpenZibo commented on src/modules/mod.rs:
```diff
@@ -71,10 +84,12 @@ impl App {
             .flat_map(|module_def| match module_def {
-                ModuleDef::Single(module) => vec![self.get_module_subscription(*module)],
+                ModuleDef::Single(module) => {
+                    vec![self.get_module_subscription(module.to_owned().into())]
```
Well... there are some other places where we should fix this kind of stuff. But in this way we can avoid some `clone`

```suggestion
                    vec![self.get_module_subscription(module)]
```

MalpenZibo reviewed (commented):
Maybe I found a solution to make me happy about the change to how we handle the `ModuleName` deserialization :smile:

MalpenZibo commented:
> The dynamic modules work like this:
> 
>     * listen command: a command that when executed yields jsons containing two fields:
>       
>       * alt: what alternative it is (e.g. swaync: notification, dnd-notification, none, ...) => based on this we will adjust icons etc
>       * text: text the module wants to display (e.g. number of notifications with swaync)
>       * example swaync output: `{"text": "16", "alt": "notification", "tooltip": "16 Notifications", "class": "notification"}`
> 
> 
> And can be configured like this: The icons config follows a regex, and if that regex is true, will display the other icon. Alert is also a regex, so if there is a match, the red dot will be shown at the right top
> 
> ```
> [[CustomModule]]
> name = "swaync"
> icon = "" # default icon
> icons.'dnd.*' = "" # icon if regex of "alt" is true, multiple such options can be given
> alert = ".*notification" # to show a red dot next to it
> listenCmd = "swaync-client -swb"
> command = "swaync-client -t -sw"
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/20912162/440183590-aeb7020a-de96-4158-93c4-3c743d69e143.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDk2NzYsIm5iZiI6MTc0NjY0OTM3NiwicGF0aCI6Ii8yMDkxMjE2Mi80NDAxODM1OTAtYWViNzAyMGEtZGU5Ni00MTU4LTkzYzQtM2M3NDNkNjllMTQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDIwMjI1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBjMTgwM2E4YTBmNTQ0YzU0ZDAyN2U5YmViMzEzYzZkMjFkMjc4ZjZkMmRkM2ZjYzhkMjdiMmFmZmEwNGM2MTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9zzYKDDBXOZEYRtbaoZMHLhtfBDF_zkS2ejwgAdHuAU)
> 
> For simplicity I followed waybars json format here, which solves my #128
> 
> I think this is ready for a first review now as well

Ok, I had to check this part but again, great work, very appreciated :100:

clotodex commented:
bit swamped rn, but custom deserializer is a good shout
Will do the changes as soon as I can get to it

MalpenZibo commented:
I merged this PR https://github.com/MalpenZibo/ashell/pull/135, so probably there are some code alignments to do

clotodex commented:
> I merged this PR #135, so probably there are some code alignments to do

Rebased with main and fixed all related config errors etc

Something I noticed is that the naming of the toml modules is not uniform
system vs mediaPlayer vs CustomModules

Is there something you prefer? should all be camelCase or snake_case? Or since it is a module upper CamleCase? Just let me know your preference and I will change it accordingly

clotodex commented:
all your suggestions implemented - thanks a lot for already writing the deserializer code!

MalpenZibo commented on src/app.rs:
```diff
@@ -81,18 +83,29 @@ pub enum Message {
         || {
             let (outputs, task) = Outputs::new(config.appearance.style, config.position);
 
+            let names = config
+                .custom_modules
+                .iter()
+                .map(|o| o.name.clone())
+                .collect::<Vec<_>>();
```
```suggestion
            let custom: HashMap<String, Custom> = config
                .custom_modules
                .iter()
                .map(|o| (o.name.clone(), Custom::default()))
                .collect();

```

MalpenZibo commented on src/app.rs:
```diff
@@ -81,18 +83,29 @@ pub enum Message {
                     logger,
                     outputs,
                     app_launcher: AppLauncher,
+                    custom: names
+                        .into_iter()
+                        .map(|n| (n.clone(), Custom::default()))
+                        .collect(),
```
```suggestion
                    custom,
```

MalpenZibo commented on src/modules/custom_module.rs:
```diff
@@ -0,0 +1,223 @@
+
+                    match command {
+                        Ok(mut child) => {
+                            let mut reader = BufReader::new(child.stdout.take().unwrap()).lines();
```
```suggestion
                            if let Some(stdout) = child.stdout.take() {
                                let mut reader = BufReader::new(stdout).lines();

```

MalpenZibo commented on src/modules/custom_module.rs:
```diff
@@ -0,0 +1,223 @@
+                                println!("child status was: {}", status);
+                            });
+
+                            while let Some(line) = reader.next_line().await.unwrap() {
```
```suggestion
                                while let Some(line) = reader.next_line().await.ok().flatten() {

```

MalpenZibo commented on src/modules/mod.rs:
```diff
@@ -225,12 +231,21 @@ impl App {
     ) -> Option<(Element<Message>, Option<OnModulePress>)> {
         match module_name {
             ModuleName::AppLauncher => self.app_launcher.view(&self.config.app_launcher_cmd),
+            ModuleName::Custom(name) => {
+                let c = self.config.custom_modules.iter().find(|m| &m.name == name);
+                if let Some(mc) = c {
+                    self.custom.get(name).unwrap().view(mc)
+                } else {
+                    error!("Custom module `{}` not found", name);
+                    None
+                }
+            }
```
```suggestion
            ModuleName::Custom(name) => {
                if let Some(v) = self
                    .config
                    .custom_modules
                    .iter()
                    .find(|m| &m.name == name)
                    .and_then(|mc| self.custom.get(name).map(|cm| cm.view(mc)))
                {
                    v
                } else {
                    error!("Custom module `{}` not found", name);
                    None
                }
            }

```

MalpenZibo commented on src/modules/mod.rs:
```diff
@@ -252,9 +267,18 @@ impl App {
+    fn get_module_subscription(&self, module_name: &ModuleName) -> Option<Subscription<Message>> {
         match module_name {
             ModuleName::AppLauncher => self.app_launcher.subscription(()),
+            ModuleName::Custom(name) => {
+                let c = self.config.custom_modules.iter().find(|m| &m.name == name);
+                if let Some(mc) = c {
+                    self.custom.get(name).unwrap().subscription(mc)
+                } else {
+                    error!("Custom module `{}` not found", name);
+                    None
+                }
+            }
```
```suggestion
            ModuleName::Custom(name) => {
                if let Some(sub) = self
                    .config
                    .custom_modules
                    .iter()
                    .find(|m| &m.name == name)
                    .and_then(|mc| self.custom.get(name).map(|cm| cm.subscription(mc)))
                {
                    sub
                } else {
                    error!("Custom module `{}` not found", name);
                    None
                }
            }

```

MalpenZibo reviewed (changes requested):
Great, we're almost there.

I left you some comments, mainly related to some unwrapping.

Sorry if I'm a bit pedantic, I really appreciate your effort

clotodex commented:
Haha thank you for the review! I see a lot of "lets get a quick thing working" stuff still stuck around - on it :D

clotodex commented on src/modules/custom_module.rs:
```diff
@@ -0,0 +1,223 @@
+                        break; // Use the first match
+                    }
+                } else {
+                    error!("Invalid regex pattern in config: {}", pattern_str);
```
needed to write a newtype so that I can use the regex in the hashmap and deserialize it, but I like it (with deref the rest of the application can use it as a regex directly)

clotodex commented:
ready to go i think - sorry for making a git mess - i think resolving conflicts from github (just cargo-toml imports) pulls in main again :D

MalpenZibo reviewed (approved):
Perfect! :tada: :rocket: :+1: :100: 

If you have time, please add an entry in the changelog and update the readme with some documentation about these custom modules. 
Otherwise, in the next few days, I will proceed with the merge and try to write something about it in the release PR.

MalpenZibo commented:
> what do you think of this?

:tada:
